### PR TITLE
os/Chown

### PR DESCRIPTION
**Pull request by leongross**
Add `os.Chown()`

### CODE REVIEW

File: src/os/os_chmod_test.go [hunk 1]
```diff
@@ -28,4 +30,26 @@ func TestChmod(t *testing.T) {
...
 	checkMode(t, f.Name(), fm)
+
+}
+
+func TestChown(t *testing.T) {
+	f := newFile("TestChown", t)
+	defer Remove(f.Name())
+	defer f.Close()
+
+	if runtime.GOOS != "windows" {
```
Comment (rminnich) [inline src/os/os_chmod_test.go]: what about plan 9?

File: src/os/os_chmod_test.go [hunk 2]
```diff
@@ -28,4 +30,26 @@ func TestChmod(t *testing.T) {
...
+		if err := Chown(f.Name(), 0, 0); err != nil {
+			t.Fatalf("chown %s 0 0: %s", f.Name(), err)
```
Comment (rminnich) [inline src/os/os_chmod_test.go]: So in terms of errors like this, and maybe this project is different, but the pattern is either
Chown(%s, 0, 0): %v != nil
or 
Chown(%s, 0, 0): got %v, want nil
It's just so you have a very regular pattern for when you need to deal with logs.
This was pretty much de rigeur at google.

File: src/os/os_chmod_test.go [hunk 3]
```diff
@@ -28,4 +30,26 @@ func TestChmod(t *testing.T) {
...
+		}
+
+		fi, err := Stat(f.Name())
+		if err != nil {
+			t.Fatalf("stat %s: %s", f.Name(), err)
+		}
+
+		if fi.Sys().(*syscall.Stat_t).Uid != 0 || fi.Sys().(*syscall.Stat_t).Gid != 0 {
```
Comment (rminnich) [inline src/os/os_chmod_test.go]: I've had tests explode with complex casts like this. A more common pattern woudl be
if fi.Sys() == nil {
error
}
s, ok := fi.Sys().(*syscall.Stat_t)
if  ! ok {
bad
}
uid, gid = s.Uid, s.Gid // just useful so you're not repeating fi.Sys().(*syscall.Stat_t). everywhere

File: src/os/os_chmod_test.go [hunk 4]
```diff
@@ -28,4 +30,26 @@ func TestChmod(t *testing.T) {
 		t.Fatalf("chmod %s %#o: %s", f.Name(), fm, err)
 	}
 	checkMode(t, f.Name(), fm)
+
+}
+
+func TestChown(t *testing.T) {
+	f := newFile("TestChown", t)
```
Comment (rminnich) [inline src/os/os_chmod_test.go:53]: What you might do first is
if runtime.GOOS == "windows" || runtime.GOOS == "plan9" {
t.Skip()
}

Then you can drop the if on line 41.

**Comment by leongross** [comment]
To design meaningful test cases for `chown`, I think you would need multiple users with different `uids` and `gids` and a user who has permission to change ownership of a file to this, a super user. In my opinion, there are 2 ways to integrate this:

1. update the container image for the CI so that there are accounts with said properties
2. mocking the file system and permissions 

The problem I see with solution 2 is that the implementation of `chown` relies on the syscall `chown`, so I wouldn't know how to mock that.

I would appreciate your input on that @rminnich 

EDIT: Maybe I am mistaken but as I see it, `golang` does not do any of the described tests but checks for an [error of an invalid gid](https://github.com/golang/go/blob/master/src/syscall/syscall_linux_test.go), maybe this is the way to go here as well?

**Comment by leongross** [comment]
@deadprogram do you have an opinion on that?

**Comment by leongross** [comment]
@deadprogram

**Comment by leongross** [comment]
@rminnich

File: src/os/file_anyos.go [hunk 5]
```diff
@@ -147,6 +147,23 @@ func Chmod(name string, mode FileMode) error {
 	return nil
 }
 
+// Chown changes the numeric uid and gid of the named file.
+// If the file is a symbolic link, it changes the uid and gid of the link's target.
+// A uid or gid of -1 means to not change that value.
+// If there is an error, it will be of type *PathError.
+//
+// On Windows or Plan 9, Chown always returns the syscall.EWINDOWS or
+// EPLAN9 error, wrapped in *PathError.
```
Comment (aykevl) [inline src/os/file_anyos.go]: TinyGo doesn't support Plan 9, so I'm not sure how useful this comment is?

**Review by aykevl — COMMENTED** [review]
I think it's fine to keep this lightly tested, for example by checking that the syscall returns an expected permission error (or even just "file not found" - it also means the call works at least somewhat).

Yes it would be great to be able to test everything, but that's very hard to do for things like this and you'll end up putting a lot of effort into what's essentially just a wrapper for a libc function.

File: src/os/file_anyos.go (hunk 5 above)
Comment (leongross) [reply src/os/file_anyos.go]: You're right, this should be removed.

**Comment by deadprogram** [comment]
@leongross can you please rebase this PR against the latest `dev`? Thanks!

**Comment by leongross** [comment]
@deadprogram rebase done, but noq quite sure where the runtime panic originates. Do you know anything about this?

**Comment by leongross** [comment]
@deadprogram @aykevl

**Comment by deadprogram** [comment]
@leongross looks like some CI fails on this PR. You probably also need to rebase against latest `dev`.

**Comment by leongross** [comment]
@deadprogram after rebasing against the current dev branch the test still pancis:

```
panic: runtime error at 0x0000000104659d7c: type assert failed
FAIL	os	0.019s
```

**Comment by leongross** [comment]
The call stack I could back trace so far looks like this:

* runtime panic
* src/runtime/interface.go:87
* compiler/interface.go:768
* compiler/compiler.go:2468

**Comment by leongross** [comment]
The old errors should be fixed except for the ci (failing probably due to https://github.com/tinygo-org/tinygo/issues/4347). Could you review it? @deadprogram

**Comment by leongross** [comment]
@deadprogram I think this was the last fix now, mind to review?

**Comment by deadprogram** [comment]
Thanks for the additions @leongross and too @rminnich and @aykevl for review.

Now squash/merging.

**Pull request merged**